src/components/splat.js

```diff
@@ -0,0 +1,910 @@
+  /* global Worker, Blob, fetch */
+  import { registerComponent } from '../core/component.js';
+  import THREE from '../lib/three.js';
+
+  registerComponent('splat', {
+    schema: {
+      src: {type: 'asset'},
+      cutoutEntity: {type: 'selector'},
+      pixelRatio: {type: 'number', default: 1},
+      xrPixelRatio: {type: 'number', default: 0.5}
+    },
+
+    init: function () {
+      this.initWorker();
+    },
+
+    update: function () {
+      var data = this.data;
+      var sceneEl = this.el.sceneEl;
+      if (data.src) { this.loadSplat(); }
+      if (data.pixelRatio > 0) {
+        sceneEl.renderer.setPixelRatio(this.data.pixelRatio);
+      }
+      if (data.xrPixelRatio > 0) {
+        sceneEl.renderer.xr.setFramebufferScaleFactor(this.data.xrPixelRatio);
+      }
+    },
+
+    loadSplat: function () {
+      var sceneEl = this.el.sceneEl;
+      var src = this.data.src;
+
+      if (!src) { return; }
+
+      // Postpone if scene not loaded.
+      if (!sceneEl.hasLoaded) {
+        sceneEl.addEventListener('loaded', this.loadSplat.bind(this));
+        return;
+      }
+
+      if (this.data.cutoutEntity) {
+        this.worldToCutoutMatrix = new THREE.Matrix4();
+        this.cutout = this.data.cutoutEntity.object3D;
+      }
+
+      var isPly = src.endsWith('.ply');
+      this.loadedVerticesNumber = 0;
+      this.worker.postMessage({method: 'clear'});
+      if (isPly) {
+        this.loadPlyFile(src);
+      } else {
+        this.loadSplatFile(src);
+      }
+    },
+
+    loadPlyFile: async function loadPlyFile (src) {
+      var response = await fetch(src);
+      var reader = response.body.getReader();
+
+      var rowLength = 3 * 4 + 3 * 4 + 4 + 4;
+      var bytesDownloaded = 0;
+      var bytesFileTotal = response.headers.get('Content-Length');
+      bytesFileTotal = bytesFileTotal ? parseInt(bytesFileTotal) : undefined;
+      var chunks = [];
+      var start = Date.now();
+      var lastReportedProgress = 0;
+
+      while (true) {
+        try {
+          var dataReceived = await reader.read();
+          if (dataReceived.done) {
+            console.log('Completed download.');
```

Comment:
I'd prefer no logging from components, but if we do it should be using `debug`.

Reply:
yeah. I'll remove. still WIP.